## 1d. Change the identity of "X"

In this example we will change the chemical identity of "X" using the Pvsat calculator, but this can be done in the same way for any calculation type in VolFe.

- "X" is an unreactive melt species whose identity can be changed. 
- The chemical identity of "X" is decided by its molecular mass, fugacity coefficient, and solubility constant.
- Current options for "X" in VolFe are Ar and Ne in basalt in rhyolite. 
- By default, "X" is Ar in basalt.

## Setting things up
You need to install VolFe once on your machine, if you haven't yet. Then we need to import a few Python packages (including VolFe). 

In [2]:
# Install VolFe on your machine.
# pip install VolFe # Remove the first # in this line if you have not installed VolFe on your machine before.

# import python packages
import pandas as pd
import VolFe as vf

For this example, we'll use analysis TN273-01D-01-01 from Brounce et al. (2014), with a temperature chosen as 1200 'C but add 20 ppm "X".

In [3]:
# Define the melt composition, fO2 estimate, and T as a dictionary.
my_analysis = {'Sample':'TN273-01D-01-01',
           'T_C': 1200., # Temperature in 'C
           'SiO2': 56.98, # wt%
           'TiO2': 1.66, # wt%
           'Al2O3': 15.52, # wt%
           'FeOT': 9.47, # wt%
           'MnO': 0.24, # wt%
           'MgO': 2.96, # wt%
           'CaO': 6.49, # wt%
           'Na2O': 4.06, # wt%
           'K2O': 0.38, # wt%
           'P2O5': 0.22, # wt%
           'H2O': 1.88, # wt%
           'CO2ppm': 13., # ppm
           'STppm': 362.83, # ppm
           'Xppm': 20., # ppm *** 20 ppm "X" added**
           'Fe3FeT': 0.171}

# Turn the dictionary into a pandas dataframe, setting the index to 0.
my_analysis = pd.DataFrame(my_analysis, index=[0])

# Show the DataFrame.
print(my_analysis)

            Sample     T_C   SiO2  TiO2  Al2O3  FeOT   MnO   MgO   CaO  Na2O  \
0  TN273-01D-01-01  1200.0  56.98  1.66  15.52  9.47  0.24  2.96  6.49  4.06   

    K2O  P2O5   H2O  CO2ppm   STppm  Xppm  Fe3FeT  
0  0.38  0.22  1.88    13.0  362.83  20.0   0.171  


By default, "X" is Ar in basalt.

In [4]:
# runs the calculation
vf.calc_Pvsat(my_analysis)

0           sample     T_C       P_bar  SiO2_wtpc TiO2_wtpc Al2O3_wtpc  \
0  TN273-01D-01-01  1200.0  587.298035  57.038397  1.661701  15.535906   

0 FeOT_wtpc  MnO_wtpc  MgO_wtpc  CaO_wtpc  ... KHOSg opt  KOSg opt KOSg2 opt  \
0  9.479706  0.240246  2.963034  6.496651  ...  Ohmoto97  Ohmoto97  ONeill22   

0  KCOg opt KCOHg opt     KOCSg opt    KCOs opt carbonylsulfide opt  \
0  Ohmoto97  Ohmoto97  Moussallam19  Holloway92                 COS   

0 density opt                        Date  
0    DensityX  2024-07-12 09:34:49.664130  

[1 rows x 173 columns]

We change this to Ar in rhyolite...

In [7]:
# choose the options I want - everything else will use the default options
my_models = [['species X solubility','Ar_Rhyolite_HughesIP']]

# turn to dataframe with correct column headers and indexes    
my_models = vf.make_df_and_add_model_defaults(my_models)

# runs the calculation
vf.calc_Pvsat(my_analysis,models=my_models)

0           sample     T_C       P_bar  SiO2_wtpc TiO2_wtpc Al2O3_wtpc  \
0  TN273-01D-01-01  1200.0  383.137402  57.038397  1.661701  15.535906   

0 FeOT_wtpc  MnO_wtpc  MgO_wtpc  CaO_wtpc  ... KHOSg opt  KOSg opt KOSg2 opt  \
0  9.479706  0.240246  2.963034  6.496651  ...  Ohmoto97  Ohmoto97  ONeill22   

0  KCOg opt KCOHg opt     KOCSg opt    KCOs opt carbonylsulfide opt  \
0  Ohmoto97  Ohmoto97  Moussallam19  Holloway92                 COS   

0 density opt                        Date  
0    DensityX  2024-07-12 09:36:21.430416  

[1 rows x 173 columns]

And Pvsat changed from 587 bar (basalt) to 383 bar (rhyolite).

Instead we can change it to Ne in basalt or rhyolite:

In [8]:
# choose the options I want - everything else will use the default options
my_models = [['species X','Ne'],['species X solubility','Ne_Basalt_HughesIP']]

# turn to dataframe with correct column headers and indexes    
my_models = vf.make_df_and_add_model_defaults(my_models)

# runs the calculation
vf.calc_Pvsat(my_analysis,models=my_models)

0           sample     T_C      P_bar  SiO2_wtpc TiO2_wtpc Al2O3_wtpc  \
0  TN273-01D-01-01  1200.0  470.38802  57.038397  1.661701  15.535906   

0 FeOT_wtpc  MnO_wtpc  MgO_wtpc  CaO_wtpc  ... KHOSg opt  KOSg opt KOSg2 opt  \
0  9.479706  0.240246  2.963034  6.496651  ...  Ohmoto97  Ohmoto97  ONeill22   

0  KCOg opt KCOHg opt     KOCSg opt    KCOs opt carbonylsulfide opt  \
0  Ohmoto97  Ohmoto97  Moussallam19  Holloway92                 COS   

0 density opt                        Date  
0    DensityX  2024-07-12 09:37:51.255225  

[1 rows x 173 columns]

In [9]:
# choose the options I want - everything else will use the default options
my_models = [['species X','Ne'],['species X solubility','Ne_Rhyolite_HughesIP']]

# turn to dataframe with correct column headers and indexes    
my_models = vf.make_df_and_add_model_defaults(my_models)

# runs the calculation
vf.calc_Pvsat(my_analysis,models=my_models)

0           sample     T_C       P_bar  SiO2_wtpc TiO2_wtpc Al2O3_wtpc  \
0  TN273-01D-01-01  1200.0  361.374399  57.038397  1.661701  15.535906   

0 FeOT_wtpc  MnO_wtpc  MgO_wtpc  CaO_wtpc  ... KHOSg opt  KOSg opt KOSg2 opt  \
0  9.479706  0.240246  2.963034  6.496651  ...  Ohmoto97  Ohmoto97  ONeill22   

0  KCOg opt KCOHg opt     KOCSg opt    KCOs opt carbonylsulfide opt  \
0  Ohmoto97  Ohmoto97  Moussallam19  Holloway92                 COS   

0 density opt                        Date  
0    DensityX  2024-07-12 09:38:17.368906  

[1 rows x 173 columns]

Note that you can mix and match the 'species X' and 'species X solubility' so that you have the molecular mass for Ar with the solubility constant of Ne in basalt... so be careful! In the future the fugacity coefficient could also be changed, but currently they can only be treated as ideal gases.